pallets/swaps/src/lib.rs

```diff
@@ -0,0 +1,317 @@
+// Copyright 2024 Centrifuge Foundation (centrifuge.io).
+// This file is part of Centrifuge Chain project.
+
+// Centrifuge is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version (see http://www.gnu.org/licenses).
+
+// Centrifuge is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+//! # Swaps pallet: Enables applying swaps independently of previous swaps in the same or opposite
+//! directions.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+pub use pallet::*;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use cfg_traits::{
+		OrderRatio, StatusNotificationHook, Swap, SwapState, SwapStatus, Swaps, TokenSwaps,
+	};
+	use frame_support::pallet_prelude::*;
+	use sp_runtime::traits::{AtLeast32BitUnsigned, EnsureAdd, EnsureSub, Zero};
+	use sp_std::cmp::Ordering;
+
+	use super::*;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	/// Configure the pallet by specifying the parameters and types on which it
+	/// depends.
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// Represents an amount that can be swapped
+		type Balance: Parameter + Member + AtLeast32BitUnsigned + Default + Copy + MaxEncodedLen;
+
+		/// An identification for a swap
+		type SwapId: Parameter + Member + Copy + Ord + MaxEncodedLen;
+
+		/// An identification for an order
+		type OrderId: Parameter + Member + Copy + Ord + MaxEncodedLen;
+
+		/// The currency type of transferrable tokens
+		type CurrencyId: Parameter + Member + Copy + MaxEncodedLen;
+
+		/// The type which exposes token swap order functionality
+		type OrderBook: TokenSwaps<
+			Self::AccountId,
+			CurrencyId = Self::CurrencyId,
+			BalanceIn = Self::Balance,
+			BalanceOut = Self::Balance,
+			OrderId = Self::OrderId,
+		>;
+
+		/// The hook which acts upon a (partially) fulfilled the swap
+		type FulfilledSwap: StatusNotificationHook<
+			Id = (Self::AccountId, Self::SwapId),
+			Status = SwapState<Self::Balance, Self::Balance, Self::CurrencyId>,
+			Error = DispatchError,
+		>;
+	}
+
+	/// Maps a `OrderId` to its corresponding `AccountId` and `SwapId`
+	///
+	/// NOTE: The storage is killed when the swap order no longer exists
+	#[pallet::storage]
+	pub(super) type OrderIdToSwapId<T: Config> =
+		StorageMap<_, Blake2_128Concat, T::OrderId, (T::AccountId, T::SwapId)>;
```

Comment:
At least this storage needs to get populated, right?

Reply:
The previous storage does not exist yet in `altair` and `centrifuge`. Just Dev, but I guess we will reset Dev with the new RU.

Reply:
But `FI` already exists, right? And we are changing that quite a lot. Don't we need a migration there, and if so, shouldn't this migration be adapted to the changes made here? Maybe @wischli can help me? If no migration needed not CR from my side.

Reply:
But the big storage refactor happened in the previous merged PR. I mean, in case of a need for a migration, we should have done that already. IIRC, we said that we could purge the previous FI (?)

Anyways, whether or not migration occurs, this PR is independent of that.

Reply:
IIRC, the point was that if there are no pending investments at the RU moment, we can reset the storage